src/secret_sharing/malicious_replicated.rs

```diff
@@ -140,38 +144,45 @@ impl<F: Field> Mul<F> for MaliciousReplicated<F> {
     }
 }
 
+#[async_trait]
 impl<F: Field> Downgrade for MaliciousReplicated<F> {
     type Target = Replicated<F>;
-    fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target>
+    {
         UnauthorizedDowngradeWrapper(self.x)
     }
 }
 
+#[async_trait]
 impl<T, U> Downgrade for (T, U)
 where
-    T: Downgrade,
-    U: Downgrade,
+    T: Downgrade + Send,
+    U: Downgrade + Send,
 {
-    type Target = (<T as Downgrade>::Target, <U as Downgrade>::Target);
-    fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
-        UnauthorizedDowngradeWrapper((
-            self.0.downgrade().access_without_downgrade(),
-            self.1.downgrade().access_without_downgrade(),
-        ))
+    type Target = (<T>::Target, <U>::Target);
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target>
+    {
+        let one = self.0.downgrade();
+        let two = self.1.downgrade();
+        let output = join(one, two).await;
+        UnauthorizedDowngradeWrapper((output.0.access_without_downgrade(), output.1.access_without_downgrade()))
```

Comment:
Even this version does not work:
let __self: (T, U)
future cannot be sent between threads safely
within `impl futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>>`, the trait `std::marker::Send` is not implemented for `<T as Downgrade>::Target`
required for the cast from `impl futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>>` to the object type `dyn futures::Future<Output = UnauthorizedDowngradeWrapper<(<T as Downgrade>::Target, <U as Downgrade>::Target)>> + std::marker::Send`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message?3#file%3A%2F%2F%2FUsers%2Frichaj%2FProposals%2FIPA%2Fraw-ipa%2Fsrc%2Fsecret_sharing%2Fmalicious_replicated.rs)
malicious_replicated.rs(167, 36): future is not `Send` as this value is used across an await
malicious_replicated.rs(167, 42): `join(one, two)` is later dropped here